Add cloud permission document for manual deployment guide for Azure (#134)
scalar-labs/scalar-kubernetes: docs/CloudPermissions.md

> # Restricting cloud privileges for deploying Scalar DL

Reviewer: It's not for restricting. It's for assigning required permissions.

Author: Changed heading to `Assign Required Cloud Permissions For Deploying Scalar DL` in c601da6

> # Restricting cloud privileges for deploying Scalar DL

Reviewer: Is it a permission or privilege?
The file name says permission and it says privilege. 
Make it consistent. 
It is described in the guide. Please don't create a PR that doesn't follow it.
https://github.com/scalar-labs/team-handbook/blob/master/engineering/guidelines/commit.md#pr-policy

Author: Sorry, It is Permission. Changed heading in c601da6

This guide explains how to restrict privileges for deploying Scalar DL on Azure.

> In general from security perspective, it is better to assign restricted privileges to users not to make them able to do operations that they are not supposed to do.

Reviewer 2: ```suggestion
 From the security perspective, It is better to give users limited privileges to protect the environment from unwanted activities.
```

Reviewer 2: You can change the suggestion in your own way.

Author: Thank you for the suggestion, updated sentence in c601da6

## Azure

> The following JSON is a custom role that allow users to manage resources for Scalar DL. This is sufficient but not necessary.

Reviewer: for Scalar DL?
or for k8s?
With the permissions, scalar dl is the only one that can be deployed? or other containers can be deployed?
if the latter, then it sounds misleading.

Author: For managing cloud resources for Scalar DL, changed the sentence in c601da6

> The following JSON is a custom role that allow users to manage resources for Scalar DL. This is sufficient but not necessary.

Reviewer 2: Sorry, What you mean by `This is sufficient but not necessary`?
Can we create without these permissions?

Author: >What you mean by This is sufficient but not necessary?

I meant this is sufficient permission but it's not necessary that the user should follow this custom role they can use an escalated permission if they want such as the `Contributor` role. Updated sentence as `These are the minimum permissions for creating a Scalar DL environment on the Azure cloud.` in c601da6

> In Azure Portal, you can create the role in Subscriptions section. Choose your subscription and select Access control (IAM) from the menu, then click `+Add` and select `Add custom role`. 

Reviewer 2: What about specifying it with bullet points?

Author: >What about specifying it with bullet points?

Thank you for the suggestion.
I used [Scalar-terraform cloud permission document ](https://github.com/scalar-labs/scalar-terraform/blob/master/docs/CloudPrivileges.md) as a reference to create this document. Should I change this?

Once the role is created, you can assign it to a user or a group from `Add role assignment` in the `+Add` menu.

> Please keep your subscription ID in the `assignableScopes` array. You can give `rolename` and `description` as per your convenience.

Reviewer: `keep` sounds like a wrong wording.

Author: Changed it to `replace` in c601da6

> Please keep your subscription ID in the `assignableScopes` array. You can give `rolename` and `description` as per your convenience.

Reviewer: rolename or roleName?

Author: Sorry, its roleName changes accordingly in c601da6

        ],
        "permissions": [
            {
                "actions": [
                    "Microsoft.Authorization/roleAssignments/*",
                    "Microsoft.Compute/availabilitySets/*",
                    "Microsoft.Compute/disks/*",
                    "Microsoft.Compute/virtualMachines/*",
                    "Microsoft.Network/loadBalancers/*",
                    "Microsoft.Network/networkInterfaces/*",
                    "Microsoft.Network/networkSecurityGroups/*",
                    "Microsoft.Network/publicIPAddresses/*",
                    "Microsoft.Network/virtualNetworks/*",
                    "Microsoft.Resources/deployments/*",
                    "Microsoft.Resources/subscriptions/resourceGroups/*",
                    "Microsoft.DocumentDb/databaseAccounts/*",
                    "Microsoft.ContainerService/managedClusters/*",
                    "Microsoft.OperationsManagement/solutions/*",
                    "Microsoft.operationalInsights/workspaces/*"
                ],
                "notActions": [],
                "dataActions": [],
                "notDataActions": []
            }
        ]
    }
}
```

> ### A Role for creating AKS cluster

Reviewer: This section talks about a role? 
It doesn't look like it. Please fix it.

Author: Changed heading to `Create Privileges for AKS cluster` in c601da6

To interact with Azure APIs, an AKS cluster requires either an Azure Active Directory (AD) service principal or a managed identity.
To create service principal for the AKS cluster ask an account with `Owner` role to create a service principal on the basis of [Azure official guide](https://docs.microsoft.com/en-us/azure/aks/kubernetes-service-principal?tabs=azure-cli). 

> Also, add `Directory Reader` permission for your user on the basis of [Azure official guide]( https://docs.microsoft.com/en-us/azure/active-directory/roles/manage-roles-portal).

Reviewer: Please explain why it needs to be added.

Author: Added explanation in c601da6

> # Assign Required Cloud Permissions For Deploying Scalar DL

Reviewer: The permissions are mainly or fully for k8s, right? 
mainly? or fully?

In any cases, why don't you say it's k8s related in the title and the first sentence?
Otherwise, it's very misleading.

Author: >The permissions are mainly or fully for k8s, right?

The permissions are for managing all the resources that are needed for Scalar DL deployment, specified that in the first sentence.

Reviewer: The question is `what do you mean Scalar DL?`
When deploying Scalar DL in k8s, Scalar DL components like scalar-ledger and scalar-envoy are just pods, so I think they don't require any permissions if k8s is properly created, do they?
In other words, if we want to deploy nginx pods and a LoadBalancer in front, then what permissions are needed? (the same permissions are required?)
So, I think the current description is too blur and somewhat wrong.

> This guide explains how to assign required permissions for the users for deploying Scalar DL on Azure.

Reviewer: Is this guide for k8s? or for scalar dl?
If you say scalar dl, then it will mean, users would need the permissions when they deploy scalar dl with terraform.
Is it correct?

Author: The permissions are for creating and managing all the resources that are needed for Scalar DL deployment. 
Specified in Azure section that the command line tools also can be used with these permissions.

> From the security perspective, it is better to give users limited privileges to protect the environment from unwanted activities.

Reviewer: privileges
make it consistent.

Author: Sorry, Changed it to permissions, changed all privileges to permissions in 6239a04.

## Azure

> The following JSON is a custom role that allows users to manage Azure cloud resources for Scalar DL. These are the minimum permissions for creating a Scalar DL environment on the Azure cloud.

Reviewer: Ditto.

Author: The permissions are for creating and managing all the resources that are needed for Scalar DL deployment.
Modified the sentence with the command line tools also can be used with these permissions.

In Azure Portal, you can create the role in the Subscriptions section. Choose your subscription and select Access control (IAM) from the menu, then click `+Add` and select `Add custom role`. 
Once the role is created, you can assign it to a user or a group from `Add role assignment` in the `+Add` menu.

Please replace your subscription ID in the `assignableScopes` array. You can give `roleName` and `description` at your convenience.

```json
{   
 "properties": {
        "roleName": "manual-guide-test",
        "description": "manual-guide-test",
        "assignableScopes": [
            "/subscriptions/xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
        ],

>         "permissions": [

Reviewer: Is this assuming that users will create a k8s cluster from the web console?
What if users us the command line tools to create k8s cluster? It is still valid?
We don't specify how to create a cluster so if this permission doc assumes the way, it is a problem.

Author: > What if users us the command line tools to create k8s cluster? It is still valid?

Yes, Users can use command line tools also. Added a sentence for it in 6239a04

            {
                "actions": [

>                     "Microsoft.Authorization/roleAssignments/*",

Reviewer: BTW, how do you derive the permissions?
Are they really necessary and sufficient?
or they are just sufficient? or maybe too sufficient?
I think we should talk about it.

Author: They are sufficient. Added sentence in Azure section.

Reviewer: Can you specify how sufficient it is?
Enterprise uses basically want minimum required permissions, so if it is very sufficient, then it is not acceptable.

                    "Microsoft.Compute/availabilitySets/*",
                    "Microsoft.Compute/disks/*",
                    "Microsoft.Compute/virtualMachines/*",
                    "Microsoft.Network/loadBalancers/*",
                    "Microsoft.Network/networkInterfaces/*",
                    "Microsoft.Network/networkSecurityGroups/*",
                    "Microsoft.Network/publicIPAddresses/*",
                    "Microsoft.Network/virtualNetworks/*",
                    "Microsoft.Resources/deployments/*",
                    "Microsoft.Resources/subscriptions/resourceGroups/*",
                    "Microsoft.DocumentDb/databaseAccounts/*",
                    "Microsoft.ContainerService/managedClusters/*",
                    "Microsoft.OperationsManagement/solutions/*",
                    "Microsoft.operationalInsights/workspaces/*"
                ],
                "notActions": [],
                "dataActions": [],
                "notDataActions": []
            }
        ]
    }
}
```

> ### Create Privileges for AKS cluster

Reviewer: privileges.
Please don't fix what you are requested to fix.

Author: Sorry, Changed it to permissions, changed all `privileges` to `permissions` in 6239a04.

> ### Create Privileges for AKS cluster

Reviewer: It suddenly says `AKS` here. Users have no idea how it is related to AKS so far.
Reorganize the whole doc once again.
First think and then organize the doc. Please don't start writing a doc before thinking about the design.
(The same as coding. We don't write code before designing the system)

Author: Description is given in the section on what is this section about and why it is needed.

# Assign Required Cloud Permissions For Deploying Scalar DL

> 

Reviewer: Also, this doc assumes that users will create an environment based on the manual guideline but it never mentions it, so it is very misleading. So the title and the descriptions should be updated.

Author: Updated the heading and sentence in 0aedc65

This guide explains how to assign required permissions for the users to create cloud resources for deploying Scalar DL on Azure.

From the security perspective, it is better to give users limited permissions to protect the environment from unwanted activities.

## Azure

The following JSON is a custom role that allows users to manage Azure cloud resources from the Azure portal or from command line tools for deploying Scalar DL.

> Please note that it is sufficient but not necessary since, it can/should be further restricted.

Reviewer: Then I think it should be more restricted, as you say...
If you let users do, then it defeats the purpose of the doc.

Reviewer: For this, it can be done later on. For now, it's fine since we don't have time.

Author: OK.

In Azure Portal, you can create the role in the Subscriptions section. Choose your subscription and select Access control (IAM) from the menu, then click `+Add` and select `Add custom role`. 
Once the role is created, you can assign it to a user or a group from `Add role assignment` in the `+Add` menu.

> Please replace your subscription ID in the `assignableScopes` array. You can give `roleName` and `description` at your convenience.

Reviewer: I think it's a wrong usage of `at your convenience`.

Author: Changed it to `as per your convenience`

```json
{   
 "properties": {

>         "roleName": "manual-guide-test",

Reviewer: ```suggestion
        "roleName": "Scalar DL permissions",
```
Don't put something meaningless.

Author: Thank you for the suggestion updated in 0aedc65

>         "description": "manual-guide-test",

Reviewer: ```suggestion
        "description": "Permissions to deploy Scalar DL on AKS",
```
Don't put something meaningless.

Author: Sorry, updated in 0aedc65

                    "Microsoft.operationalInsights/workspaces/*"
                ],
                "notActions": [],
                "dataActions": [],
                "notDataActions": []
            }
        ]
    }
}
```

Note:

Following are the permissions for each assignment

* `Microsoft.Authorization/roleAssignments/*` : To assign roles for the resources.
* `Microsoft.Compute/availabilitySets/*` : To create and manage availabilitySets.
* `Microsoft.Compute/disks/*` : To create and manage disks for virtual machines.
* `Microsoft.Network/loadBalancers/*` : To create and manage Load balancers.
* `Microsoft.Network/networkInterfaces/*` : To create and manage network interfaces
* `Microsoft.Network/networkSecurityGroups/*` : To create and manage security groups
* `Microsoft.Network/publicIPAddresses/*` : To create and manage public IP address.
* `Microsoft.Network/virtualNetworks/*` : To create and manage virtual network.
* `Microsoft.Resources/deployments/*` : To validate the Azure deployment template for creating the resources.
* `Microsoft.Resources/subscriptions/resourceGroups/*` : To create and manage resource groups.

> * `Microsoft.DocumentDb/databaseAccounts/*`: To create and manage Cosmos DB.

Reviewer: I think it's too specific.
We need to separate this.
like, in CosmosDB case, add xxx, for another database, add yyy or something.

Author: Created a table and separated it.
